comp/remoteagent/impl/remoteagent.go

```diff
@@ -0,0 +1,337 @@
+		shutdownChan: shutdownChan,
+	}
+
+	deps.Lifecycle.Append(fx.Hook{
+		OnStart: func(context.Context) error {
+			go comp.Start()
+			return nil
+		},
+		OnStop: func(context.Context) error {
+			shutdownChan <- struct{}{}
+			return nil
+		}})
+
+	return comp
+}
+
+// RemoteAgent is the main registry for remote agents. It tracks which remote agents are currently registered, when
+// they were last seen, and handles collecting status and flare data from them on request.
+type RemoteAgent struct {
+	agentMap     map[string]*remoteAgentDetails
+	agentMapMu   sync.Mutex
+	shutdownChan chan struct{}
+}
+
+// RegisterRemoteAgent registers a remote agent with the registry.
+//
+// If the remote agent is not present in the registry, it is added. If a remote agent with the same ID is already
+// present, the API endpoint and display name are checked: if they are the same, then the "last seen" time of the remote
+// agent is updated, otherwise the remote agent is removed and replaced with the incoming one.
+func (ra *RemoteAgent) RegisterRemoteAgent(registration *remoteagent.RegistrationData) (uint32, error) {
+	ra.agentMapMu.Lock()
+	defer ra.agentMapMu.Unlock()
+
+	// If the remote agent is already registered, then we might be dealing with an update (i.e., periodic check-in) or a
+	// brand new remote agent. As the agent ID may very well not be a unique ID every single time (it could always just
+	// be `process-agent` or what have you), we differentiate between the two scenarios by checking the human name and
+	// API endpoint give to us.
+	//
+	// If either of them are different, then we remove the old remote agent and add the new one. If they're the same,
+	// then we just update the last seen time and move on.
+	id := util.SanitizeAgentID(registration.ID)
+	entry, ok := ra.agentMap[id]
+	if !ok {
+		// We've got a brand new remote agent, so do a sanity check by trying to connect to their gRPC endpoint and if
+		// that works, add them to the registry.
+		details, err := newRemoteAgentDetails(registration)
+		if err != nil {
+			return 0, err
+		}
+
+		log.Infof("Remote agent '%s' registered.", id)
+		ra.agentMap[id] = details
+		return remoteAgentRecommendedRefreshIntervalSecond, nil
+	}
+
+	// We already have an entry for this remote agent, so check if we need to update the gRPC client, and then update
+	// the other bits.
+	if entry.apiEndpoint != registration.APIEndpoint {
+		entry.apiEndpoint = registration.APIEndpoint
+
+		client, err := newRemoteAgentClient(registration)
+		if err != nil {
+			return 0, err
+		}
+		entry.client = client
+	}
+
+	entry.displayName = registration.DisplayName
+	entry.lastSeen = time.Now()
+
+	return remoteAgentRecommendedRefreshIntervalSecond, nil
+}
+
+// Start starts the remote agent registry, which periodically checks for idle remote agents and deregisters them.
+func (ra *RemoteAgent) Start() {
+	go func() {
+		log.Info("Remote Agent registry started.")
+
+		ticker := time.NewTicker(5 * time.Second)
```

Comment:
```suggestion
		ticker := time.NewTicker(remoteAgentQueryTimeout)
```

Reply:
Also, would it make sense to allow to tweak this value with a configuration value?

Reply:
I think it might make sense to allow tweaking how long we let remote agents sit around in an idle state for, but I'm not sure it's a big deal to change how often we check if any remote agents have exceeded the allowable idle TTL, which is what the hard-coded 5 seconds is for here.

Reply:
I moved the idle timeout, query timeout, and recommended refresh interval to be configurable.. but left this particular interval fixed at five seconds, based on the above explanation.

Let me know if you still think it should be configurable.